src/parthenon_array_generic.hpp

```diff
@@ -0,0 +1,339 @@
+using all_true = std::is_same<bool_pack<bs..., true>, bool_pack<true, bs...>>;
+
+template <class... Ts>
+using are_all_integral = all_true<std::is_integral<Ts>::value...>;
+
+template <class T>
+bool all_greater_than(T) {
+  return true;
+}
+template <class T, class Arg, class... Args>
+bool all_greater_than(T val, Arg v, Args... args) {
+  return (v > val) && all_greater_than(val, args...);
+}
+
+template <class... Ts>
+using enable_if_all_integral = std::enable_if<are_all_integral<Ts...>::value>;
+} // namespace impl
+
+using namespace impl;
+
+struct empty_state_t {
+  // Allow this to be constructed with anything
+  template <class... Args>
+  KOKKOS_INLINE_FUNCTION empty_state_t(Args &&...) {}
+
+  // LFR: We can't easily have virtual destructors on device, see Kokkos issue #1591.
+  // It is non-ideal for there not to be a virtual destructor on this and other
+  // state structs that ParArrayGeneric ends up inheriting from, but it is unlikely
+  // that there will be a need to use these with runtime polymorphism.
+  // KOKKOS_INLINE_FUNCTION
+  // virtual ~empty_state_t() {};
+};
+
+// API designed with Data = Kokkos::View<T******> in mind
+template <typename Data, typename State = empty_state_t>
+class ParArrayGeneric : public State {
+ public:
+  using index_pair_t = std::pair<size_t, size_t>;
+  using base_t = Data;
+  using state_t = State;
+  using HostMirror = ParArrayGeneric<typename Data::HostMirror, State>;
+  using host_mirror_type = HostMirror;
+
+  ParArrayGeneric() = default;
+  __attribute__((nothrow)) ~ParArrayGeneric() = default;
+  __attribute__((nothrow))
+  ParArrayGeneric(const ParArrayGeneric<Data, State> &t) = default;
+  __attribute__((nothrow)) ParArrayGeneric<Data, State> &
+  operator=(const ParArrayGeneric<Data, State> &t) = default;
+  __attribute__((nothrow)) ParArrayGeneric(ParArrayGeneric<Data, State> &&t) = default;
+  __attribute__((nothrow)) ParArrayGeneric<Data, State> &
+  operator=(ParArrayGeneric<Data, State> &&t) = default;
+
+  KOKKOS_INLINE_FUNCTION
+  explicit ParArrayGeneric(const Data &v, const State &state = State())
+      : State(state), data_(v) {}
+
+  template <class Data2, class State2>
+  KOKKOS_INLINE_FUNCTION explicit ParArrayGeneric(
+      const ParArrayGeneric<Data2, State2> &arr_in)
+      : State(static_cast<State2>(arr_in)), data_(arr_in.data_) {}
+
+  template <class State2, class... Ts>
+  operator ParArrayGeneric<Kokkos::View<Ts...>, State2>() const {
+    return ParArrayGeneric<Kokkos::View<Ts...>, State2>(data_, static_cast<State>(*this));
+  }
+
+  // Allow a ParArrayGeneric to be cast to any compatible Kokkos view
+  template <class... Ts>
+  operator Kokkos::View<Ts...>() const {
+    return data_;
+  }
+
+  template <class... Args, class = typename enable_if_all_integral<Args...>::type>
+  ParArrayGeneric(const std::string &label, Args... args)
+      : ParArrayGeneric(label, State(),
+                        std::make_index_sequence<Data::rank - sizeof...(Args)>{},
```

Comment:
Yeah, it turns out to have a pretty wide range of uses.

Reply:
Yes, it produces `0 1 2 ... N-1` but you can use the comma operator to unpack an `index_sequence` to a list of the same size but just containing one value repeated, i.e. 
```c++ 
template <std::size_t... I>
std::vector<int> GetVectorOfOnes(std::index_sequence<I...>) {
      retrun {(void) I, 1)...};
}
```
returns a vector of `1`s with a size equal to the size of the index sequence. This is essentially the main trick used to make everything generic. The `(void)` cast is just to suppress compiler warnings about unused variables.

Reply:
I see. Would you mind adding a brief comment to the first use of `make_index_sequence` and first use of ` ((void)I, 1)` to explain this?
I think this could be helpful for other developers looking at this code in the future as this might not be obvious to everyone (including me :smile: )

Reply:
Sure, I will add a comment.

Reply:
done.